### ISSUE TITLE
C# Extract method selects wrong part of code

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#58971]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/C-Extract-method-selects-wrong-part-of-/1637567)._

---
[regression] [worked-in:17.1 preview 2]
Reproduction:
1. Open a C# code file
2. Select something which can be extracted to a method
3. Press Ctrl+R, Ctrl+M to launch the Extract Method command
4. The method is extracted and entered the rename phase.
5. The selection is wrong so we can't just type in the new method name.

Please see the following screenshot (it should have selected "NewMethod", however, it selected "wMethod()").

![image](https://user-images.githubusercontent.com/1408396/151230909-575f3844-cf18-47f8-95dd-cec5d8c02b87.png)

---
#### Feedback Bot on 1/17/2022, 00:42 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by KirillOsenkov** [comment]
Seeing this a lot, the resulting selection seems to be off:

Extract method for this selection:
![image](https://user-images.githubusercontent.com/679326/158688309-f49ba62b-5d9d-4cc7-bcb3-bd443549225a.png)

results in:
![image](https://user-images.githubusercontent.com/679326/158688195-e552dc5f-7f00-48e9-96e1-8a3baf6fdcaf.png)

**Comment by KirillOsenkov** [comment]
This happens when the initial selection overlaps with some preceding whitespace in the trivia. If you select the statement without the trivia, it works fine.

**Comment by BhaaLseN** [comment]
My 2ct (was I was about to open another issue, but then saw this one):

There is rarely ever the case that I only select a single line (or part of a line) and extract it into a method. When I do that, it is at least two, often many more lines. And I certainly don't (and won't) take the time to start the selection in the first line at the first character just to satisfy the whitespace condition there.

What I do is just select the full lines using the editor margin - and I'm guessing many folks do the same.

My expectation would even be that the full selection is considered, moved to the extracted method; and _both_ locations are formatted/indented according to the current style settings (either `.editorconfig` or VS), with the full method name being a linked identifier to change its name.

**Comment by CyrusNajmabadi** [comment]
@jinnhoward I can't repro this. Can you give q shark example that does this for you?  Thanks!

**Comment by BhaaLseN** [comment]
I can reproduce on 17.2.1 with this:
```c#
public class Testing
{
    public void ExtractHere()
    {
        int x = 42;
        int mul = x * x;
        Console.WriteLine(mul);
    }
}
```
Select the two lines _after_ `int x = 42;` then hit the refactoring hotkey (in my case: Ctrl+R, M)
![extract-method](https://user-images.githubusercontent.com/826052/168899730-c882e7f9-9d51-45cd-afbf-1aa6e2f8646f.gif)

**Comment by CyrusNajmabadi** [comment]
THanks.  I'll try that out.

**Comment by CyrusNajmabadi** [comment]
Does not repro for me.  Will upload gif momentarily.

**Comment by CyrusNajmabadi** [comment]
![NoRepro](https://user-images.githubusercontent.com/4564579/168904968-ffa765c1-f408-46d5-b955-2f85fead6dc3.gif)

**Comment by CyrusNajmabadi** [comment]
Oh... i see.  It's not the lightbulb.  It's the explicit command.  Yes.  this repros for me now.  Will investigate.

**Comment by CyrusNajmabadi** [comment]
Ok.  Found issue and have fix available.

**Comment by jinnHoward** [comment]
Awesome! Can't wait. Thanks!

**Comment by maxima120** [comment]
+1 (btw - its been broken for a while)

**Comment by CyrusNajmabadi** [comment]
@maxima120 this has been fixed.

**Comment by maxima120** [comment]
it is still an issue in 17.1.6

**Comment by ryzngard** [comment]
The fix is expected to ship in the 17.3 time frame

**Comment by toreaurstadboss** [comment]
This should have been fixed sooner. Not good. Wasting people work hours on manually doing essential refactoring stuff. Who does not use the extract method feature and how could such an essential functionality be deployed into a production version and why have we waited for months on this ?

**Comment by CyrusNajmabadi** [comment]
@toreaurstadboss PRs always welcome :)

> Who does not use the extract method feature 

Personally, i just use the lightbulb (which did not have this issue).  It was only an issue if you explicitly went through teh command handler (which appears to be much less used).  So it was missed.

**Comment by toreaurstadboss** [comment]
Well, many users hit just Ctrl+"." to see quick fix actions or hit Ctrl+R+m - maybe it is me just old school using the GUI lol

**Comment by maxima120** [comment]
it's not you. It's the environment. There are now a lot more fantasist developers than actual developers in the world and they all have big millennial mouths and tons of time to go around and complain (cos they aren't employed).

**Comment by CyrusNajmabadi** [comment]
> maybe it is me just old school using not the GUI but keyboard shortcuts lol

Sure.  My point is simply, if things are not receiving barely any reports, they will get less attention from things with many reports.  So please let us know when you hit things so we can feed those signals into triaging process, and we can ensure we're addressing problems in a reasonable order.

**Comment by ryzngard** [comment]
From my understanding this shipped in 17.3 and is fixed. I tested using mouse only as well as keyboard selection for command. Please let me know if you're still hitting this in latest RTM builds
